clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java

```diff
@@ -469,6 +472,333 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+                    consumer.seek(partition, 0);
+            }
+        });
+
+        // lookup coordinator
+        client.prepareResponseFrom(new GroupCoordinatorResponse(Errors.NONE.code(), node).toStruct(), node);
+
+        Node coordinator = new Node(Integer.MAX_VALUE - node.id(), node.host(), node.port());
+
+        // join group
+        client.prepareResponseFrom(joinGroupFollowerResponse(assignor, 1, "memberId", "leaderId", Errors.NONE.code()), coordinator);
+
+        // sync group
+        client.prepareResponseFrom(syncGroupResponse(Arrays.asList(partition), Errors.NONE.code()), coordinator);
+
+        consumer.poll(0);
+
+        // respond to the outstanding fetch so that we have data available on the next poll
+        client.respondFrom(fetchResponse(partition, 0, 5), node);
+        client.poll(0, time.milliseconds());
+
+        // the next poll should return the completed fetch
+        consumer.poll(0);
+
+        // 5 records have been returned to the consumer and auto commit is disabled, so no commit offset request should be sent
+        consumer.unsubscribe();
+
+        for (ClientRequest req: client.requests()) {
+            assertTrue(req.request().header().apiKey() != ApiKeys.OFFSET_COMMIT.id);
+        }
+    }
+
+    /**
+     * "consumer" subscribes to "topic1" and "topic2" (each with only one partition) and has auto-commit enabled.
+     * "consumer" reads 5 records from the single partition of "topic1" from offset 0.
+     * "consumer" changes subscription to "topic2" only.
+     * Since auto-commit is enabled, the last committed offset of "topic1" partition should be 5 and of "topic2" partition should be 0.
+     */
+    @Test
+    public void testSubscriptionChangeAndOffsetCommitWithAutoCommitEnabled() {
+        String topic1 = "topic1";
+        final TopicPartition partition1 = new TopicPartition(topic1, 0);
+        String topic2 = "topic2";
+        final TopicPartition partition2 = new TopicPartition(topic2, 0);
+    
+        int sessionTimeoutMs = 30000;
+        int heartbeatIntervalMs = 3000;
+    
+        // adjust auto commit interval lower than heartbeat so we don't need to deal with
+        // a concurrent heartbeat request
+        int autoCommitIntervalMs = 1000;
+    
+        Time time = new MockTime();
+        MockClient client = new MockClient(time);
+        Cluster cluster = TestUtils.singletonCluster(topic1, 1);
+        Node node = cluster.nodes().get(0);
+        client.setNode(node);
+        Metadata metadata = new Metadata(0, Long.MAX_VALUE);
+        metadata.update(cluster, time.milliseconds());
+        PartitionAssignor assignor = new RoundRobinAssignor();
+    
+        final KafkaConsumer<String, String> consumer = newConsumer(time, client, metadata, assignor,
+                sessionTimeoutMs, heartbeatIntervalMs, true, autoCommitIntervalMs);
+        consumer.subscribe(Arrays.asList(topic1, topic2), new ConsumerRebalanceListener() {
```

Comment:
Minor: it seems like this test would be a little simpler if you start subscribed to topic1 and then switch to topic2.